tests/components/pilight/test_init.py

```diff
@@ -66,351 +61,340 @@ def set_callback(self, function):
-    @patch("homeassistant.components.pilight._LOGGER.error")
-    def test_send_code_delay(self, mock_pilight_error):
-        """Try to send proper data with delay afterwards."""
-        with assert_setup_component(4):
-            assert setup_component(
-                self.hass,
-                pilight.DOMAIN,
-                {pilight.DOMAIN: {pilight.CONF_SEND_DELAY: 5.0}},
+            mock_client.assert_called_once_with(
+                host=pilight.DEFAULT_HOST, port=pilight.DEFAULT_PORT
             )
+            assert mock_error.call_count == 1
 
-            # Call with protocol info, should not give error
-            service_data1 = {"protocol": "test11", "value": 42}
-            service_data2 = {"protocol": "test22", "value": 42}
-            self.hass.services.call(
+
+@patch("pilight.pilight.Client", PilightDaemonSim)
+async def test_send_code_no_protocol(hass):
+    """Try to send data without protocol information, should give error."""
+    with assert_setup_component(4):
+        assert await async_setup_component(hass, pilight.DOMAIN, {pilight.DOMAIN: {}})
+
+        # Call without protocol info, should raise an error
+        try:
+            await hass.services.async_call(
                 pilight.DOMAIN,
                 pilight.SERVICE_NAME,
-                service_data=service_data1,
+                service_data={"noprotocol": "test", "value": 42},
                 blocking=True,
             )
-            self.hass.services.call(
+            await hass.async_block_till_done()
+        except MultipleInvalid as error:
+            assert "required key not provided @ data['protocol']" in str(error)
+
+
+@patch("homeassistant.components.pilight._LOGGER.error")
+@patch("homeassistant.components.pilight._LOGGER", _LOGGER)
+@patch("pilight.pilight.Client", PilightDaemonSim)
+async def test_send_code(mock_pilight_error, hass):
+    """Try to send proper data."""
+    with assert_setup_component(4):
+        assert await async_setup_component(hass, pilight.DOMAIN, {pilight.DOMAIN: {}})
+
+        # Call with protocol info, should not give error
+        service_data = {"protocol": "test", "value": 42}
+        await hass.services.async_call(
+            pilight.DOMAIN,
+            pilight.SERVICE_NAME,
+            service_data=service_data,
+            blocking=True,
+        )
+        await hass.async_block_till_done()
+        error_log_call = mock_pilight_error.call_args_list[-1]
+        service_data["protocol"] = [service_data["protocol"]]
+        assert str(service_data) in str(error_log_call)
+
+
+@patch("pilight.pilight.Client", PilightDaemonSim)
+@patch("homeassistant.components.pilight._LOGGER.error")
+async def test_send_code_fail(mock_pilight_error, hass):
+    """Check IOError exception error message."""
+    with assert_setup_component(4):
+        with patch("pilight.pilight.Client.send_code", side_effect=IOError):
+            assert await async_setup_component(
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6372789)